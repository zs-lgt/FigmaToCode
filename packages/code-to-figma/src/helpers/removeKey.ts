export const removeKey = node => {
    if (node && node.key) {
        delete node.key;
    }
    if (node && node.children) {
        node.children.forEach(removeKey);
    }
    return node;
};
