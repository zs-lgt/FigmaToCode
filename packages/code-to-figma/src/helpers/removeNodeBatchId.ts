export const removeNodeBatchId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.nodeBatchId;
    }
    if (node && node.children) {
        node.children.forEach(removeNodeBatchId);
    }
    return node;
};
