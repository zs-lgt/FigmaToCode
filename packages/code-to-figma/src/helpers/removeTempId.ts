export const removeTempId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.reactId;
    }
    if (node && node.children) {
        node.children.forEach(removeTempId);
    }
    return node;
};
