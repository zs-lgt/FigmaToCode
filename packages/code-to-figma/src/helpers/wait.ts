import { act } from 'react-test-renderer';

export function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}
