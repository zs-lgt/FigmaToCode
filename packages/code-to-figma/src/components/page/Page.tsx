import * as React from 'react';
import { BaseNodeProps, ChildrenProps, ExportProps, StyleOf } from '../../types';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { $currentPageTempId, api } from '../../rpc';
import { map } from 'rxjs/operators';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { InheritStyleProvider } from '../../hooks/useInheritStyle';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { StyleSheet } from '../../helpers/StyleSheet';
import { useImageHash } from '../../hooks/useImageHash';

export interface PageProps extends BaseNodeProps, ChildrenProps, ExportProps, OnLayoutHandlerProps {
    style?: StyleOf<YogaStyleProperties & GeometryStyleProperties>;
    isCurrent?: boolean;
    onCurrentChange?: (isCurrent: boolean) => void;
    backgrounds?: ReadonlyArray<Paint>;
}

export const useCurrentPageChange = (
    nodeRef: { current?: any },
    callback?: ((isCurrent: boolean) => void) | void,
    _isCurrent?: boolean
) => {
    React.useEffect(() => {
        const instance = nodeRef.current;
        const subject = $currentPageTempId.pipe(map((message: any) => message === instance.reactId));

        const subscription = subject.subscribe(value => callback && callback(value));

        return () => subscription.unsubscribe();
    }, []);

    React.useEffect(() => {
        if (_isCurrent) {
            api.setCurrentPage(nodeRef.current);
        }
    }, [_isCurrent]);
};

const Page: React.FC<PageProps> = props => {
    const nodeRef = React.useRef();
    const { onCurrentChange, ...otherProps } = props;
    useCurrentPageChange(nodeRef, onCurrentChange, props.isCurrent);

    const yogaChildProps = useYogaLayout({ nodeRef, ...otherProps });
    useOnLayoutHandler(yogaChildProps, props);

    const style = StyleSheet.flatten(props.style);

    const imageHash = useImageHash(style.backgroundImage);

    const pageProps = {
        ...transformGeometryStyleProperties('backgrounds', style, imageHash),
        ...otherProps,
        style
    };

    return (
        <InheritStyleProvider style={props.style}>
            <page {...pageProps} {...yogaChildProps} innerRef={nodeRef} />
        </InheritStyleProvider>
    );
};

export { Page };
