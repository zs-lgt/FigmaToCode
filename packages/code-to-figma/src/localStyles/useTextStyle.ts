import { api } from '../rpc';
import { Platform } from '../helpers/Platform';
import { TextStyleProperties, transformTextStyleProperties } from '../styleTransformers/transformTextStyleProperties';
import * as React from 'react';
import { useFontName } from '../hooks/useFontName';
import { CommonStyleProps } from '../types';
import { useCreateFillStyleId } from './useCreateFillStyleId';

const AVAILABLE_TEXT_PROPERTIES = [
    'fontSize',
    'textDecoration',
    'fontName',
    'letterSpacing',
    'lineHeight',
    'paragraphIndent',
    'paragraphSpacing',
    'textCase'
];

export const useTextStyle = (style: Partial<TextStyleProperties>, params: CommonStyleProps) => {
    const [textStyleId, setTextStyleId] = React.useState(null);

    const transformedStyles = React.useMemo(
        () => ({
            ...transformTextStyleProperties(style),
            style
        }),
        [style]
    );

    const [createFillsStyle, fillStyleId] = useCreateFillStyleId(
        style && style.color ? transformedStyles.fills : null,
        params
    );

    const textProperties = React.useMemo(() => {
        return Object.keys(transformedStyles).reduce(
            (prev, current) => ({
                ...prev,
                ...(AVAILABLE_TEXT_PROPERTIES.indexOf(current) >= 0 ? { [current]: transformedStyles[current] } : {})
            }),
            {}
        );
    }, [transformedStyles]);

    const loadedFont = useFontName((textProperties as any).fontName || { family: 'Roboto', style: 'Regular' });

    React.useEffect(() => {
        if (Platform.OS !== 'figma') {
            return;
        }
        const createTextStyle = async () => {
            if (Object.keys(textProperties).length === 0) {
                return;
            }
            const id = await api.createOrUpdateTextStyle({
                textProperties,
                params,
                loadedFont
            });
            setTextStyleId(id);
        };

        if (loadedFont) {
            createTextStyle();
        }

        createFillsStyle();
    }, [textProperties, loadedFont]);

    return {
        ...style,
        ...(textStyleId ? { textStyleId: textStyleId } : {}),
        ...(fillStyleId ? { fillStyleId: fillStyleId } : {})
    };
};
