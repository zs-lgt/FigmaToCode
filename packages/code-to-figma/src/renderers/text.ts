import { baseNodeMixin } from '../mixins/baseNodeMixin';
import { geometryMixin } from '../mixins/geometryMixin';
import { layoutMixin } from '../mixins/layoutMixin';
import { saveStyleMixin } from '../mixins/saveStyleMixin';
import { propsAssign } from '../helpers/propsAssign';
import { exportMixin } from '../mixins/exportMixin';
import { TextProps } from '../components/text/Text';
import { blendMixin } from '../mixins/blendMixin';
import { isValidSize } from '../helpers/isValidSize';
import { isEqualFontStyle } from '../helpers/isEqualFontStyle';
import { sceneNodeMixin } from '../mixins/sceneNodeMixin';
import { uiApi } from '../rpc';
import { safeGetPluginData } from '../helpers/safeGetPluginData';
import { constraintsMixin } from '../mixins/constraintsMixin';
import { DEFAULT_FONT } from '../helpers/constants';

const textNodePropsAssign = propsAssign<TextProps, TextProps>(
    [
        'characters',
        'textAlignHorizontal',
        'textAlignVertical',
        'paragraphIndent',
        'paragraphSpacing',
        'autoRename',
        'fontSize',
        'textCase',
        'textDecoration',
        'letterSpacing',
        'lineHeight',
        'textStyleId',
        'hyperlink'
    ],
    {
        characters: '',
        textAlignHorizontal: 'LEFT',
        textAlignVertical: 'TOP',
        paragraphIndent: 0,
        paragraphSpacing: 0,
        autoRename: false,
        fontSize: 12,
        textCase: 'ORIGINAL',
        textDecoration: 'NONE',
        letterSpacing: { value: 0, unit: 'PIXELS' },
        lineHeight: { unit: 'AUTO' },
        hyperlink: null
    }
);

const defaultFont = DEFAULT_FONT;

export const text = (node: TextNode) => (props: TextProps & { loadedFont?: FontName; hasDefinedWidth?: boolean }) => {
    const textNode = node || props.node || figma.createText();

    baseNodeMixin(textNode)(props);
    saveStyleMixin(textNode)(props);
    layoutMixin(textNode)(props);
    geometryMixin(textNode)(props);
    exportMixin(textNode)(props);
    blendMixin(textNode)(props);
    sceneNodeMixin(textNode)(props);
    constraintsMixin(textNode)(props);

    const { loadedFont, fontName = defaultFont } = props;
    if (
        loadedFont &&
        fontName &&
        loadedFont.family === fontName.family &&
        isEqualFontStyle(loadedFont.style, fontName.style)
    ) {
        if (props.fontName) {
            textNode.fontName = loadedFont;
        }
        if (
            props.hasDefinedWidth &&
            isValidSize(props.width) &&
            isValidSize(textNode.height) &&
            !props.textAutoResize
        ) {
            textNode.resize(props.width, textNode.height);
            textNode.textAutoResize = 'HEIGHT';
        } else {
            textNode.textAutoResize = props.textAutoResize || 'WIDTH_AND_HEIGHT';
        }

        const oldCharacters = textNode.characters;
        const oldFontSize = textNode.fontSize;
        textNodePropsAssign(textNode)(props);
        if (oldCharacters !== textNode.characters || oldFontSize !== textNode.fontSize) {
            const reactId = safeGetPluginData('reactId')(textNode);
            if (reactId) {
                uiApi.updateYogaNode(reactId);
            }
        }
    }

    return textNode;
};
