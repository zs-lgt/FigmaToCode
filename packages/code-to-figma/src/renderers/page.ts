import { baseNodeMixin } from '../mixins/baseNodeMixin';
import { saveStyleMixin } from '../mixins/saveStyleMixin';
import { exportMixin } from '../mixins/exportMixin';
import { pageMixin } from '../mixins/pageMixin';
import { PageProps } from '../components/page/Page';

export const page = (node: PageNode) => (props: PageProps) => {
    const page = node || figma.createPage();

    saveStyleMixin(page)(props);
    baseNodeMixin(page)(props);
    exportMixin(page)(props);

    pageMixin(page)(props);

    return page;
};
