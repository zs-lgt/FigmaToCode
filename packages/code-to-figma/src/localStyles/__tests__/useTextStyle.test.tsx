import * as React from 'react';
import { createFigma } from 'figma-api-stub';
import { render } from '../../renderer';
import { wait } from '../../helpers/wait';
import { Text } from '../../components/text/Text';
import { Page } from '../../components/page/Page';
import { removeNodeBatchId } from '../../helpers/removeNodeBatchId';
import { removeTempId } from '../../helpers/removeTempId';

const removeMeta = node => {
    return removeKey(removeNodeBatchId(removeTempId(node)));
};

jest.mock('nanoid');
import { nanoid } from 'nanoid';
import { useTextStyle } from '../useTextStyle';
import { removeKey } from '../../helpers/removeKey';

describe('useTextStyle', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('useTextStyle correctly applied', async () => {
        // @ts-ignore
        nanoid.mockReturnValue('test');
        const Component = () => {
            const rootStrokeStyle = useTextStyle(
                {
                    fontFamily: 'Roboto',
                    fontWeight: 'bold',
                    fontSize: 24,
                    color: '#000000'
                } as any,
                {
                    name: 'root/background'
                }
            );
            return <Text style={rootStrokeStyle} />;
        };
        await render(
            <Page>
                <Component />
            </Page>
        );
        await wait();
        expect(removeMeta(figma.root)).toMatchSnapshot();
        expect(figma.getLocalTextStyles()).toMatchSnapshot();
    });
});
