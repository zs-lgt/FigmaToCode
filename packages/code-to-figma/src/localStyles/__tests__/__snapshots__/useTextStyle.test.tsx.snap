// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTextStyle useTextStyle correctly applied 1`] = `
ChildrenMixinStub {
  "children": Array [
    ExportMixinStub {
      "children": Array [],
      "config": undefined,
      "id": "0:1",
      "parent": [Circular],
      "type": "PAGE",
    },
    ExportMixinStub {
      "_backgrounds": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "b": 0.9607843160629272,
            "g": 0.9607843160629272,
            "r": 0.9607843160629272,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "children": Array [
        GeometryMixinStub {
          "_characters": "",
          "_fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0,
                "r": 0,
              },
              "opacity": 1,
              "type": "SOLID",
            },
          ],
          "_fontName": Object {
            "family": "Roboto",
            "style": "Bold",
          },
          "_textAutoResize": "WIDTH_AND_HEIGHT",
          "autoRename": false,
          "blendMode": "NORMAL",
          "config": Object {
            "isWithoutTimeout": false,
            "simulateErrors": true,
          },
          "constraints": Object {
            "horizontal": "MIN",
            "vertical": "MIN",
          },
          "effects": Array [],
          "exportSettings": Array [],
          "fillStyleId": "S:test,",
          "fontSize": 24,
          "hyperlink": null,
          "id": "1:2",
          "isMask": false,
          "layoutAlign": "INHERIT",
          "layoutGrow": 0,
          "letterSpacing": Object {
            "unit": "PIXELS",
            "value": 0,
          },
          "lineHeight": Object {
            "unit": "AUTO",
          },
          "locked": false,
          "opacity": 1,
          "paragraphIndent": 0,
          "paragraphSpacing": 0,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"fontFamily\\":\\"Roboto\\",\\"fontWeight\\":\\"bold\\",\\"fontSize\\":24,\\"color\\":\\"#000000\\",\\"textStyleId\\":\\"S:test,\\",\\"fillStyleId\\":\\"S:test,\\"}",
          },
          "strokeAlign": "INSIDE",
          "strokeCap": "NONE",
          "strokeWeight": 0,
          "strokes": Array [],
          "textAlignHorizontal": "LEFT",
          "textAlignVertical": "TOP",
          "textCase": "ORIGINAL",
          "textDecoration": "NONE",
          "textStyleId": "S:test,",
          "type": "TEXT",
          "visible": true,
        },
      ],
      "config": Object {
        "isWithoutTimeout": false,
        "simulateErrors": true,
      },
      "exportSettings": Array [],
      "id": "1:1",
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "config": undefined,
  "id": "0:0",
  "type": "DOCUMENT",
}
`;

exports[`useTextStyle useTextStyle correctly applied 2`] = `
Array [
  TextStyleStub {
    "fontName": Object {
      "family": "Roboto",
      "style": "Bold",
    },
    "fontSize": 24,
    "id": "S:test,",
    "name": "root/background",
    "remote": false,
    "styleBasics": Object {
      "effectStyles": Array [],
      "gridStyles": Array [],
      "paintStyles": Array [
        PaintStyleStub {
          "id": "S:test,",
          "name": "root/background",
          "paints": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0,
                "r": 0,
              },
              "opacity": 1,
              "type": "SOLID",
            },
          ],
          "remote": false,
          "styleBasics": [Circular],
          "type": "PAINT",
        },
      ],
      "styles": Map {
        "S:test," => [Circular],
      },
      "textStyles": [Circular],
    },
    "type": "TEXT",
  },
]
`;
